src/main/scala/is/hail/io/LoadMatrix.scala

```diff
@@ -0,0 +1,231 @@
+      if (dropSamples)
+        Array.empty
+      else
+        header1
+
+    val nSamples = sampleIds.length
+
+    LoadMatrix.warnDuplicates(sampleIds)
+
+    val lines = sc.textFilesLines(files, nPartitions.getOrElse(sc.defaultMinPartitions))
+
+    val fileByPartition = lines.partitions.map(p => partitionPath(p))
+    val firstPartitions = fileByPartition.zipWithIndex.filter((name) => name._2 == 0 || fileByPartition(name._2-1) != name._1).map((name) => name._2)
+
+    val matrixType = MatrixType(VSMMetadata(
+      sSignature = TString,
+      vSignature = TString,
+      genotypeSignature = TInt64
+    ))
+
+    val keyType = matrixType.kType
+    val rowKeys: RDD[RegionValue] = lines.mapPartitionsWithIndex { (i,it) =>
+
+      if (firstPartitions.contains(i)) {
+        val hd1 = header1Bc.value
+        val hd = it.next().value.split(sep)
+        if (!hd1.sameElements(hd)) {
+          hd1.zipAll(hd, None, None).zipWithIndex.foreach { case ((s1, s2), j) =>
+            if (s1 != s2) {
+              fatal(
+                s"""invalid sample ids: expected sample ids to be identical for all inputs. Found different sample ids at position $j.
+                   |    ${files(0)}: $s1
+                   |    ${fileByPartition(i)}: $s2""".
+                  stripMargin)
+            }
+          }
+        }
+      }
+
+      val region = MemoryBuffer()
+      val rvb = new RegionValueBuilder(region)
+      val rv = RegionValue(region)
+
+      new Iterator[RegionValue] {
+        var present = false
+
+        def advance() {
+          while (!present && it.hasNext) {
+            it.next().foreach { line =>
+              if (line.nonEmpty) {
+                val k = line.substring(0,line.indexOf(sep))
+                region.clear()
+                rvb.start(keyType)
+                rvb.startStruct()
+                rvb.addString(k)
+                rvb.addString(k)
+                rvb.endStruct()
+                rv.setOffset(rvb.end())
+
+                present = true
+              }
+            }
+          }
+        }
+
+        def hasNext: Boolean = {
+          if (!present)
+            advance()
+          present
+        }
+
+        def next(): RegionValue = {
+          hasNext
+          assert(present)
+          present = false
+          rv
+        }
+      }
+    }
+
+    val rdd = lines
+      .mapPartitionsWithIndex { (i,it) =>
+
+        val region = MemoryBuffer()
+        val rvb = new RegionValueBuilder(region)
+        val rv = RegionValue(region, 0)
+
+        if (firstPartitions.contains(i)) {it.next()}
+
+        new Iterator[RegionValue] {
+          var present = false
+
+          def advance() {
+            while (!present && it.hasNext) {
+              it.next().foreach { line =>
+                if (line.nonEmpty) {
+
+                  val firstsep = line.indexOf(sep)
+
+                  region.clear()
+                  rvb.start(matrixType.orderedRDD2Type.rowType.fundamentalType)
```

Comment:
You don't need `.fundamentalType`.  If that's still in LoadVCF, it should get cleaned up.  (I'll check.)